microsoft/vscode
# Extend support for data breakpoints in VS Code API

## What it does

The breakpoint API in VS Code supports to add

- Allow the user to manage data breakpoints through vscode.debug API
- Add methods to gain information about a potential data breakpoint
- Allow data breakpoints to have different origins
  - Keep use case where data id is already known (current)
  - Add infrastructure for already existing address resolution

Communication:
- Adapt DataBreakpoint with source between extension host and main
- Expose DataBreakpoint in VS Code API, previously not exposed (should not cause any backwards compatibility issues)

Minor:
- Make bytes optional in data bytes info, as it is the same in the DAP
- Make access type optional in data breakpoint, as it is the same in the DAP

Fixes https://github.com/microsoft/vscode/issues/195151

## How to test

To test this change, you need to make sure you have a C++ file that you can debug as the Typescript debugger does not support data breakpoints as far as I can see.
1. Install [VS Code C++ Tools](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools)
2. Create a C++ file, e.g.,

```cpp
#include <iostream>

int main() {
    int counter = 0;

    for (int i = 0; i < 10; i++) {
        counter++;
        std::cout << "Counter: " << counter << std::endl;
    }

    return 0;
}
```

3. Create and start an extension (or modify an existing one that is already started with VS Code) with the code similar to this one:
```ts
	vscode.debug.onDidChangeBreakpoints(evt => {
		vscode.window.showInformationMessage(JSON.stringify(evt, undefined, 2));
	});

	vscode.debug.onDidChangeActiveDebugSession(session => {
		setTimeout(async () => {
			if (!session) {
				return;
			}
			try {
				const info = await new vscode.VariableScopedDataBreakpointOrigin(1000, 'i').info(session);
				vscode.window.showInformationMessage(JSON.stringify(info, undefined, 2));
				if (info.dataId) {
					vscode.debug.addBreakpoints([new vscode.DataBreakpoint(info as vscode.AvailableDataBreakpointInfo)]);
				}
			} catch (error) {
				vscode.window.showErrorMessage(error.message);
			}
			try {
				const info = await new vscode.AddressDataBreakpointOrigin('0x7fffffffd7e8', 4).info(session);
				vscode.window.showInformationMessage(JSON.stringify(info, undefined, 2));
			} catch (error) {
				vscode.window.showErrorMessage(error.message);
			}
		}, 2000);
	});
	vscode.window.showInformationMessage('Breakpoints:', vscode.debug.breakpoints.length.toString());

	vscode.debug.addBreakpoints([
		new vscode.DataBreakpoint(new vscode.ResolvedDataBreakpointOrigin('0x7fffffffd7e8,counter,4'), 'readWrite', false, 'condition', 'hitCondition', 'logMessage'),
		new vscode.DataBreakpoint(new vscode.AddressDataBreakpointOrigin('0x7fffffffd7e8', 4), 'readWrite', false, 'condition', 'hitCondition', 'logMessage'),
		new vscode.DataBreakpoint(new vscode.VariableScopedDataBreakpointOrigin(1000, 'i'), 'readWrite', false, 'condition', 'hitCondition', 'logMessage'),
		new vscode.FunctionBreakpoint('Hello')
	]);
```

When starting the extension you should already see the three data breakpoints that you added:
![image](https://github.com/user-attachments/assets/da528ed2-346c-47ae-8eb0-9a60b7598d1c)

4. Enable the data breakpoints (or already produce them enabled in your extension code).
5. Debug the program with the `Debug C/C++ File` command.
6. Depending on your other setup not all breakpoints may resolve correctly but you can see that they are properly evaluated.
